Deprecate `containerPackageInfo` parameter in `createOdspCreateContainerRequest`

MarioJGMsoft opened the pull request:
## Description

As of PR https://github.com/microsoft/FluidFramework/pull/22849 (released in 2.23), preferred pattern to add "containerPackageName" parameter to request is via URL resolver instead of `createOdspCreateContainerRequest`'s last parameter. 
Tag calling `createOdspCreateContainerRequest` with that parameter as deprecated now, and then in the 2.40 release remove that form altogether.

## Future Breaking Changes

See [issue 23882](https://github.com/microsoft/FluidFramework/issues/23882) for more details.

## Reviewer Guidance

Please let me know if there's anything else I should be aware of or can do better, thanks!

Fixes: [AB#31450](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/31450)

MarioJGMsoft committed:
refactor: created overload so that we can properly deprecate the parameter containerPackageInfo

MarioJGMsoft committed:
refactor: had to disable import no deprecated.

MarioJGMsoft commented on examples/utils/webpack-fluid-loader/src/odspUrlResolver.ts:
```diff
@@ -11,6 +11,7 @@ import {
 } from "@fluidframework/odsp-doclib-utils/internal";
 import {
 	OdspDriverUrlResolver,
+	// eslint-disable-next-line import/no-deprecated
```
Had to disable import no deprecated, because even though the version that is being imported is technically not the deprecated version, eslint detected that it was using that one. Does anyone know if there's a better way to deprecate the parameter?

jason-ha replied:
The linter doesn't understand / is being conservative. This is okay, but should add a comment with the work item that will clean this up.

jason-ha replied:
I think the only alternative might be to import the module as in `import * as OdspDriver from "...`.
But since it is temporary, this is good enough.

jason-ha replied:
Still looking for comment with the work item number.

MarioJGMsoft replied:
Added comment

jason-ha replied:
I recently learned of a better format to comment lint disables using `--`:
```typescript
	// eslint-disable-next-line import/no-deprecated -- This is the non-deprecated call, but eslint doesn't distinguish. TODO AB#31049 remove suppression
```

MarioJGMsoft committed:
Merge branch 'main' into deprecateRPI

jason-ha commented on examples/utils/webpack-fluid-loader/src/odspUrlResolver.ts:
```diff
@@ -11,6 +11,7 @@ import {
 } from "@fluidframework/odsp-doclib-utils/internal";
 import {
 	OdspDriverUrlResolver,
+	// eslint-disable-next-line import/no-deprecated
 	createOdspCreateContainerRequest,
 	createOdspUrl,
 } from "@fluidframework/odsp-driver/internal";
```
This is an example and should not be importing from /internal. Recommend fixing these when we have to touch them. See examples that use `examples/.eslintrc.cjs` and/or `examples/.eslintrc.data.cjs`.

markfields replied:
cc @Abe27342 and @alexvy86 -- We were just talking about these in Foundation Enhancements meeting.  This example which does stuff with ODSP Driver probably should be in a different group of examples that are internal-only. We don't expect anyone in the public to be doing anything with ODSP Driver.

jason-ha replied:
Should none of the odsp-driver exports be /legacy then? At least some of these are and should be okay to use /legacy import instead of /internal.

jason-ha replied:
thread nudge

jason-ha replied:
more aggressive nudge (@markfields , @Abe27342 , @alexvy86 )

jason-ha commented on packages/drivers/odsp-driver/src/createOdspCreateContainerRequest.ts:
```diff
@@ -12,6 +12,36 @@ import { ISharingLinkKind } from "@fluidframework/odsp-driver-definitions/intern
 import { buildOdspShareLinkReqParams, getContainerPackageName } from "./odspUtils.js";
 
+/**
+ * Original version of createOdspCreateContainerRequest
```
`Original version` is not a useful comment. You might be able to have an ts-doc block with just `/** @deprecated */` or that and the level tags or that with inheritdoc (more suspect). I don't know if we have any examples of this -- so you'll have to play around. The ts-doc comments impact the ff.com content. You can find this one here: https://fluidframework.com/docs/api/odsp-driver/#createodspcreatecontainerrequest-function
Do a local docs (website) build and make sure the content looks good with changes.

jason-ha replied:
@Josmithr, do you already know what to expect in docs when there are overloads?

Josmithr replied:
Yeah. I was about to comment on this, but decided not to since this was mimicking an existing pattern.

Users _only_ see docs for the overload signatures. As far as I know, no one sees the docs on the implementation. So these docs are inverted. Elsewhere, we generally leave the implementation undocumented aside from the release tags.

Josmithr replied:
I just added a brief section about this to our TSDoc guidelines: https://github.com/microsoft/FluidFramework/wiki/TSDoc-Guidelines#release-tags

MarioJGMsoft replied:
I don't understand to what I should change the comment to, since I can't leave the comment block empty without dsabling eslint require-description.

MarioJGMsoft replied:
Yeah, I get the message "Single line blocks are not permitted in your configuration".

jason-ha replied:
And if you make multi-line?
I often write `/** foo */` just as shorthand to mean a ts-doc with that content in it.

MarioJGMsoft replied:
I then get the message: Missing JSDoc block description. I can't leave the @ tag by itself. In the example that was linked (https://github.com/microsoft/FluidFramework/wiki/TSDoc-Guidelines#documenting-function-with-overloads), the final function only has /** @public */, but that format is invalid because it's single line and doesn't have a description.

If you make it multiline, then it still doesn't count the tag as a description.

My question is, where should I keep the original comment block?

jason-ha replied:
Per Josh, for docs, it sounds like you can put whatever you want in there.
I would probably just duplicate comments, but as best I can tell IntelliSense also doesn't pay attention to implementation docs.
@Josmithr, probably should update the example

Josmithr commented on packages/drivers/odsp-driver/src/createOdspCreateContainerRequest.ts:
```diff
@@ -12,6 +12,36 @@ import { ISharingLinkKind } from "@fluidframework/odsp-driver-definitions/intern
+ * Overloaded version of createOdspCreateContainerRequest that takes in containerPackageInfo
+ * @legacy
+ * @alpha
+ * @deprecated - To be removed in 2.40
```
`@deprecated` comments don't want a `-` delimiter. This will result in generated docs that read "- To be removed in 2.40".

I would also recommend noting explicitly that there is a new overload of the same function that should be used instead.

Josmithr replied:
For reference: https://github.com/microsoft/FluidFramework/wiki/TSDoc-Guidelines#deprecated

MarioJGMsoft replied:
The deprecated version is the overloaded one, so how can I express this in the documentation?

jason-ha replied:
The `@deprecated` tag is incorporated into the documentation. You can also make a comment on the "deprecated" parameter's `@param` for clarity.

Josmithr reviewed (dismissed):
Left a suggestion to enhance the deprecation notice comment. Otherwise, the API changes look good.

jason-ha commented on packages/test/test-drivers/src/odspDriverApi.ts:
```diff
@@ -12,6 +12,7 @@ import {
 import {
 	OdspDocumentServiceFactory,
 	OdspDriverUrlResolver,
+	// eslint-disable-next-line import/no-deprecated
 	createOdspCreateContainerRequest,
 	createOdspUrl,
 } from "@fluidframework/odsp-driver/internal";
```
Other option for this and other /internal import is to make the /internal export only export the non-deprecated version.
The how-to is covered at https://github.com/microsoft/FluidFramework/wiki/Change-Recipes#moving-api-to-internal

The ./internal.ts version could look like:
```typescript
export const odspCreateContainerRequest: (
	siteUrl: string,
	driveId: string,
	filePath: string,
	fileName: string,
	createShareLinkType?: ISharingLinkKind,
) => ReturnType<typeof createOdspCreateContainerRequestOriginal> = createOdspCreateContainerRequestOriginal;
```

Since there are only two internal uses, may not be worth the effort.

MarioJGMsoft replied:
I think it's not worth the effort.

jason-ha commented:
I don't quite follow this statement in the description: "With PR: #22849 we introduced a parameter for ODSP that will not be used."
The parameter is used when passed. When present there will be a query parameter `containerPackageName=...`.
Maybe you mean to say that there is alternative? Or are you saying that the generated QP will later be ignored?

MarioJGMsoft commented:
> I don't quite follow this statement in the description: "With PR: #22849 we introduced a parameter for ODSP that will not be used." The parameter is used when passed. When present there will be a query parameter `containerPackageName=...`. Maybe you mean to say that there is alternative? Or are you saying that the generated QP will later be ignored?

The query parameter (I'm guessing that's what QP stands for), will later be ignored. The new alternative is that it can be added in the constructor of `OdspDriverUrlResolverForShareLink`. Then it will be processed and added when it resolves a request.

Therefore, having the parameter in the request is unnecessary.

jason-ha commented:
> > I don't quite follow this statement in the description: "With PR: #22849 we introduced a parameter for ODSP that will not be used." The parameter is used when passed. When present there will be a query parameter `containerPackageName=...`. Maybe you mean to say that there is alternative? Or are you saying that the generated QP will later be ignored?
> 
> The query parameter (I'm guessing that's what QP stands for), will later be ignored. The new alternative is that it can be added in the constructor of `OdspDriverUrlResolverForShareLink`. Then it will be processed and added when it resolves a request.
> 
> Therefore, having the parameter in the request is unnecessary.

I am not sure I understand what it exactly means to be processed and added when the request is resolved. Sounds like the deprecation issue needs some more content. If originally someone passed `containerPackageInfo` and for `IRequest.url` with `containerPackageName=...` that was then ..??.., what are all of the changes that are needed to get the same effect? Sounds like something should be changed in some resolver.?

MarioJGMsoft commented:
> > > I don't quite follow this statement in the description: "With PR: #22849 we introduced a parameter for ODSP that will not be used." The parameter is used when passed. When present there will be a query parameter `containerPackageName=...`. Maybe you mean to say that there is alternative? Or are you saying that the generated QP will later be ignored?
> > 
> > 
> > The query parameter (I'm guessing that's what QP stands for), will later be ignored. The new alternative is that it can be added in the constructor of `OdspDriverUrlResolverForShareLink`. Then it will be processed and added when it resolves a request.
> > Therefore, having the parameter in the request is unnecessary.
> 
> I am not sure I understand what it exactly means to be processed and added when the request is resolved. Sounds like the deprecation issue needs some more content. If originally someone passed `containerPackageInfo` and got `IRequest.url` with `containerPackageName=...` that was then ..??.., what are all of the changes that are needed to get the same effect? Sounds like something should be changed in some resolver.?

The original idea was that if someone passed the containerPackageName through the request, when that request would be resolved, the resolved url would contain the containerPackageName in the shareLinkInfo web url. 

After trying to integrate said solution, I realized that it would be better to have it in the constructor of the resolver, because that information is always required by the url, so instead of having to copy and paste the same information everytime a request is made, I thought it was better to have it set once in the resolver, always being passed along and stored in shareLinkInfo web url. 

The only necessary changes should be that:
1. Instead of having containerPackageInfo in the request, you have it in `OdspDriverUrlResolverForShareLink`
2. When you resolve a request expecting said information in the shareLink, you use a `OdspDriverUrlResolverForShareLink` resolver (Which people should already be doing).

You are correct that a resolver should be changed (specifically, `OdspDriverUrlResolverForShareLink`), and those changes already did happen in this PR: https://github.com/microsoft/FluidFramework/pull/23878

jason-ha commented:
> The original idea was that if someone passed the containerPackageName through the request, when that request would be resolved, the resolved url would contain the containerPackageName in the shareLinkInfo web url.
> 
> After trying to integrate said solution, I realized that it would be better to have it in the constructor of the resolver, because that information is always required by the url, so instead of having to copy and paste the same information everytime a request is made, I thought it was better to have it set once in the resolver, always being passed along and stored in shareLinkInfo web url.
> 
> The only necessary changes should be that:
> 
> 1. Instead of having containerPackageInfo in the request, you have it in `OdspDriverUrlResolverForShareLink`
> 2. When you resolve a request expecting said information in the shareLink, you use a `OdspDriverUrlResolverForShareLink` resolver (Which people should already be doing).
> 
> You are correct that a resolver should be changed (specifically, `OdspDriverUrlResolverForShareLink`), and those changes already did happen in this PR: #23878

Let's update the issue.
- The issue says "Now, we have the capability to add that information inside the resolver itself so that it is only set once and always used when a request is made and processed." But I think it should say "In version 2.xx, `OdspDriverUrlResolverForShareLink` constructor takes optional `IContainerPackageInfo` that will provide functionality more efficiently."
- There should be a note for custom resolvers. Perhaps "If using a custom resolver, update it to provide this capability similar to #23878."

MarioJGMsoft committed:
docs: added changelist

MarioJGMsoft committed:
docs: updated text format

MarioJGMsoft committed:
docs: updated comments

MarioJGMsoft committed:
docs: changed changeset doc name

MarioJGMsoft committed:
docs: updated comment

jason-ha commented on .changeset/wicked-eagles-search.md:
```diff
@@ -0,0 +1,14 @@
+"section": deprecation
+---
+
+createOdspCreateContainerRequest().containerPackageInfo is now deprecated
```
`createOdspCreateContainerRequest().containerPackageInfo` is not a thing.

MarioJGMsoft replied:
Applied change

jason-ha reviewed (commented):
Can you run this through the FF-Loop integration pipeline?
I don't know if we have any proof that deprecations are being handled properly yet.

MarioJGMsoft committed:
docs: updated changeset

jason-ha commented on .changeset/wicked-eagles-search.md:
```diff
@@ -0,0 +1,14 @@
+"section": deprecation
+---
+
+The containerPackageInfo parameter of the createOdspCreateContainerRequest() function will be deprecated and removed in 2.40.
+
+The parameter `containerPackageInfo` in `createOdspCreateContainerRequest()` is deprecated and will be removed in version 2.40.
```
Duplicate and somewhat at odds texts. While much the same one says "will be deprecated" and the other says "is deprecated". The second is correct. So just delete the first, I think.

jason-ha replied:
Still duplicated - remove lines 8 and 9.

MarioJGMsoft committed:
docs: updated changeset

MarioJGMsoft committed:
docs: updated changeset based on comments

MarioJGMsoft committed:
Merge remote-tracking branch 'origin' into deprecateRPI

MarioJGMsoft committed:
refactor: added fix for eslint deprecation problems

MarioJGMsoft committed:
docs: removed eslint no deprecated.

MarioJGMsoft committed:
Merge branch 'microsoft:main' into deprecateRPI

MarioJGMsoft committed:
Merge branch 'main' into deprecateRPI

MarioJGMsoft committed:
docs: Disabled eslint deprecation

MarioJGMsoft committed:
Merge branch 'main' into deprecateRPI

jason-ha commented on packages/test/test-drivers/src/odspTestDriver.ts:
```diff
@@ -397,12 +398,14 @@ export class OdspTestDriver implements ITestDriver {
 		return new this.api.OdspDriverUrlResolver();
 	}
 
+	sharingLinkKind = { scope: SharingLinkScope.organization };
 	createCreateNewRequest(testId: string): IRequest {
 		return this.api.createOdspCreateContainerRequest(
 			this.config.siteUrl,
 			this.config.driveId,
 			this.config.directory,
 			`${testId}.tstFluid`,
+			this.sharingLinkKind,
```
What are these changes for?
Are you trying to avoid an indication that this is a deprecated use?

MarioJGMsoft replied:
Those were test changes that I accidentally committed, will revert.

jason-ha commented on packages/drivers/odsp-driver/src/createOdspCreateContainerRequest.ts:
```diff
@@ -20,7 +20,51 @@ import { buildOdspShareLinkReqParams, getContainerPackageName } from "./odspUtil
+ * If not given that means that the container package does not have a name.
+ * @legacy
+ * @alpha
+ * @deprecated To be removed in 2.40
+ * Add containerPackageInfo to the OdspDriverUrlResolverForShareLink constructor instead; see https://github.com/microsoft/FluidFramework/issues/23882 for more details.
+ * Deprecating overloaded function to remove containerPackageInfo
+ */
+export function createOdspCreateContainerRequest(
+	siteUrl: string,
+	driveId: string,
+	filePath: string,
+	fileName: string,
+	createShareLinkType: ISharingLinkKind | undefined,
+	containerPackageInfo?: IContainerPackageInfo | undefined,
+): IRequest;
```
To avoid any confusion in tooling that is trying to decide which overload applies, the `?` should be removed from `containerPackageInfo` parameter. Then this signature should only be applicable when that parameter is actually specified.
I think this is the reason the order of declaration mattered. In some cases either signature applied and tooling could just chose the first one. I don't know if it will make any difference to eslint checks.

MarioJGMsoft replied:
Applied change

jason-ha commented on packages/drivers/odsp-driver/src/createOdspCreateContainerRequest.ts:
```diff
@@ -20,7 +20,51 @@ import { buildOdspShareLinkReqParams, getContainerPackageName } from "./odspUtil
+	createShareLinkType?: ISharingLinkKind,
+): IRequest;
+
+/**
+ * Overloaded version of createOdspCreateContainerRequest that takes in containerPackageInfo
```
Just use the same basic description.
```suggestion
/**
 * Create the request object with url and headers for creating a new file on OneDrive Sharepoint
```

MarioJGMsoft replied:
Applied change

MarioJGMsoft committed:
Merge branch 'main' into deprecateRPI

MarioJGMsoft committed:
docs: applied changes based on comments

MarioJGMsoft committed:
docs: applied changes based on comments

jzaffiro commented on .changeset/wicked-eagles-search.md:
```diff
@@ -0,0 +1,12 @@
+"section": deprecation
+---
+
+The parameter `containerPackageInfo` in `createOdspCreateContainerRequest()` is deprecated and will be removed in version 2.40.
```
```suggestion
Deprecate the `containerPackageInfo` parameter in `createOdspCreateContainerRequest()`
```

MarioJGMsoft replied:
Applied changes

jason-ha replied:
@tylerbutler and @jzaffiro - I think Mario got comments both ways on whether the removal version should be included in this text.

jzaffiro commented on .changeset/wicked-eagles-search.md:
```diff
@@ -0,0 +1,12 @@
+
+The parameter `containerPackageInfo` in `createOdspCreateContainerRequest()` is deprecated and will be removed in version 2.40.
+
+This will mean that the name of the containerPackage can no longer be sent through the request. Instead it can be added in the constructor of `OdspDriverUrlResolverForShareLink`.
```
```suggestion
The name of the containerPackage can no longer be sent through the request. This functionality will be removed in version 2.40. Instead, it can be added in the constructor of `OdspDriverUrlResolverForShareLink`.
```

MarioJGMsoft replied:
Applied changes

MarioJGMsoft committed:
docs: applied changes based on docs review

tylerbutler commented on .changeset/wicked-eagles-search.md:
```diff
@@ -0,0 +1,13 @@
+---
+"@fluidframework/odsp-driver": minor
+---
+---
+"section": deprecation
+---
```
```suggestion
---
"@fluidframework/odsp-driver": minor
"__section": deprecation
---
```

The changeset format recently changed, but since this was created before the build-tools update, we can just fix it manually.

MarioJGMsoft replied:
Applied change

tylerbutler commented on .changeset/wicked-eagles-search.md:
```diff
@@ -0,0 +1,13 @@
+"section": deprecation
+---
+
+Deprecate the `containerPackageInfo` parameter in `createOdspCreateContainerRequest()` is deprecated and will be removed in version 2.40.
```
```suggestion
The containerPackageInfo parameter in createOdspCreateContainerRequest is now deprecated
```

MarioJGMsoft replied:
Applied change

tylerbutler commented on .changeset/wicked-eagles-search.md:
```diff
@@ -0,0 +1,13 @@
+
+Deprecate the `containerPackageInfo` parameter in `createOdspCreateContainerRequest()` is deprecated and will be removed in version 2.40.
+
+The name of the containerPackage can no longer be sent through the request. This functionality will be removed in version 2.40.
```
```suggestion
The `containerPackageInfo` parameter in `createOdspCreateContainerRequest()` is deprecated will be removed in version 2.40.0.

The name of the containerPackage can no longer be sent through the request. Instead, it can be added in the constructor of `OdspDriverUrlResolverForShareLink`.
```

MarioJGMsoft replied:
Applied change

tylerbutler commented on .changeset/wicked-eagles-search.md:
```diff
@@ -0,0 +1,13 @@
+Deprecate the `containerPackageInfo` parameter in `createOdspCreateContainerRequest()` is deprecated and will be removed in version 2.40.
+
+The name of the containerPackage can no longer be sent through the request. This functionality will be removed in version 2.40.
+Instead it can be added in the constructor of `OdspDriverUrlResolverForShareLink`.
```
```suggestion
```

MarioJGMsoft replied:
Applied change

tylerbutler commented on .changeset/wicked-eagles-search.md:
```diff
@@ -0,0 +1,13 @@
+The name of the containerPackage can no longer be sent through the request. This functionality will be removed in version 2.40.
+Instead it can be added in the constructor of `OdspDriverUrlResolverForShareLink`.
+
+See [issue](https://github.com/microsoft/FluidFramework/issues/23882) for details.
```
```suggestion
See [issue #23822](https://github.com/microsoft/FluidFramework/issues/23882) for more details.
```

MarioJGMsoft replied:
Applied change

MarioJGMsoft committed:
docs: Applied changes to changeset based on comments

MarioJGMsoft committed:
docs: updated based on comments

tylerbutler commented on .changeset/wicked-eagles-search.md:
```diff
@@ -0,0 +1,12 @@
+"__section": deprecation
+---
+
+The `containerPackageInfo` parameter in `createOdspCreateContainerRequest()` is now deprecated.
```
```suggestion
The containerPackageInfo parameter in createOdspCreateContainerRequest is now deprecated
```

MarioJGMsoft replied:
Applied changes

tylerbutler commented on .changeset/wicked-eagles-search.md:
```diff
@@ -0,0 +1,12 @@
+
+The `containerPackageInfo` parameter in `createOdspCreateContainerRequest()` is now deprecated.
+
+The `containerPackageInfo` parameter in `createOdspCreateContainerRequest()` is deprecated will be removed in version 2.40.0.
```
```suggestion
The `containerPackageInfo` parameter in `createOdspCreateContainerRequest()` is deprecated and will be removed in version 2.40.0.
```

MarioJGMsoft replied:
Applied changes

MarioJGMsoft committed:
applied changes based on comments.

MarioJGMsoft committed:
Applied changes based on comments

jzaffiro reviewed (approved):
Changeset looks good to me!

MarioJGMsoft committed:
Merge branch 'main' into deprecateRPI

tylerbutler reviewed (approved):
Approving for docs.

jatgarg committed:
Merge branch 'main' into deprecateRPI

MarioJGMsoft committed:
Merge branch 'main' into deprecateRPI
